src/wallet/wallet.cpp

```diff
@@ -5141,13 +5141,19 @@ bool CWallet::ParameterInteraction()
         }
     }
 
-    // -zapwallettx implies a rescan
-    if (gArgs.GetBoolArg("-zapwallettxes", false)) {
+    int zapwallettxes = gArgs.GetArg("-zapwallettxes", 0);
+    // -zapwallettxes implies dropping the mempool on startup
+    if (zapwallettxes != 0 && gArgs.SoftSetBoolArg("-persistmempool", false)) {
```

Comment:
Need to fix `LoadMempool()` too
```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 461616ae2..789068fcd 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -4358,7 +4358,7 @@ static const uint64_t MEMPOOL_DUMP_VERSION = 1;
 
 bool LoadMempool(void)
 {
-    if (gArgs.GetBoolArg("-zapwallettxes", false)) {
+    if (gArgs.GetBoolArg("-zapwallettxes", false) && !gArgs.GetBoolArg("-persistmempool", DEFAULT_PERSIST_MEMPOOL)) {
         LogPrintf("Skipping mempool.dat because of zapwallettxes\n");
         return true;
     }
```
or just revert #2782. Should then probably also fix buttons in Repair dialog i.e. call zapwallettxes options with `persistmempool=0` (assuming that's the expected behaviour though I'm not sure about that):
```diff
diff --git a/src/qt/rpcconsole.cpp b/src/qt/rpcconsole.cpp
index 0a1712676..e74bb95bf 100644
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -57,8 +57,8 @@ const TrafficGraphData::GraphRange INITIAL_TRAFFIC_GRAPH_SETTING = TrafficGraphD
 // Repair parameters
 const QString SALVAGEWALLET("-salvagewallet");
 const QString RESCAN("-rescan");
-const QString ZAPTXES1("-zapwallettxes=1");
-const QString ZAPTXES2("-zapwallettxes=2");
+const QString ZAPTXES1("-zapwallettxes=1 -persistmempool=0");
+const QString ZAPTXES2("-zapwallettxes=2 -persistmempool=0");
 const QString UPGRADEWALLET("-upgradewallet");
 const QString REINDEX("-reindex");
 
```

Reply:
well, which diff do you think we should do?

Reply:
I would probably revert #2782 (the less code the better) and apply `rpcconsole.cpp` patch (in which case we would preserve existing behaviour of repair options for GUI users).